horovod/common/controller.cc

```diff
@@ -0,0 +1,743 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <atomic>
+#include <map>
+#include <queue>
+#include <set>
+#include <unordered_set>
+
+#include "controller.h"
+#include "global_state.h"
+#include "logging.h"
+#include "operations.h"
+
+namespace horovod {
+namespace common {
+
+// This function performs all the preparation work for workers to agree
+// on what tensors to be all-reduced or all-gathered. The output is a
+// response list that includes all tensors that are ready.
+//
+// The coordinator follows a master-worker paradigm. Rank zero acts
+// as the master (the "coordinator"), whereas all other ranks are simply
+// workers. Each worker maintains a cache of tensors that are previously
+// broadcasted as ready by other ranks. If the cache covers all incoming
+// messages, there's no need for workers to do additional communications.
+// Otherwise, workers will communicate with each other to agree on what
+// tensors to be processed. The communication performs as following:
+//
+//      a) The workers send a Request to the coordinator, indicating what
+//      they would like to do (which tensor they would like to gather and
+//      reduce, as well as their shape and type). They repeat this for every
+//      tensor that they would like to operate on.
+//
+//      b) The workers send an empty "DONE" message to the coordinator to
+//      indicate that there are no more tensors they wish to operate on.
+//
+//      c) The coordinator receives the Requests from the workers, as well
+//      as from its own TensorFlow ops, and stores them in a request table.
+//      The coordinator continues to receive Request messages until it has
+//      received GLOBAL_SIZE number of empty "DONE" messages.
+//
+//      d) The coordinator finds all tensors that are ready to be reduced,
+//      gathered, or all operations that result in an error. For each of
+//      those, it sends a Response to all the workers. When no more
+//      Responses are available, it sends a "DONE" response to the workers.
+//      If the process is being shutdown, it instead sends a "SHUTDOWN"
+//      response.
+//
+//      e) The workers listen for Response messages, processing each one by
+//      doing the required reduce or gather, until they receive a "DONE"
+//      response from the coordinator. At that point, the tick ends.
+//      If instead of "DONE" they receive "SHUTDOWN", they mark it in the
+//      response list.
+ResponseList Controller::ComputeResponseList() {
+  // Update cache capacity if autotuning is active.
+  if (global_state_.param_manager.IsAutoTuning()) {
```

Comment:
Instead of injecting global_state, can we inject the necessary components?  Or are we using more than half of the state?

Reply:
We are using nearly half of them, is there a specific concern for not directly storing the reference to global state?

Reply:
It just hides dependencies, makes the code look like it depends on "everything".

Reply:
I see. Refracted.